libs/ngxtension/mapfilter/src/mapfilter.spec.ts

```diff
@@ -0,0 +1,57 @@
+import { from } from 'rxjs';
+import { mapFilter } from './mapfilter';
+
+describe(mapFilter.name, () => {
+	it('given an observable >1-42-3| and a mapping function that double ONLY the odds value, then result is an observable of >"2"--"6"-| the intial even value (42) is not mapped and so filtered out', () => {
+		const in$ = from([1, 42, 3]);
+		const out$ = in$.pipe(
+			mapFilter((n) => {
+				if (n % 2) return String(n * 2);
+				else return undefined; // explict return undefined to skipout (filter) some value from the out observable
+			})
+		);
+
+		let out = '>';
+		out$.subscribe((s) => {
+			//s is a string
+			out += s + '-';
+		});
+		out += '|';
+		expect(out).toEqual('>2-6-|');
```

Comment:
I don't get your test. Why are you not simply testing the returned value ? It makes your test very hard to understand. 
We could add observableSpy to eaiser test observable. What do you thing @nartc ?

Reply:
@tomalaforge I'm not an test expert expecially for RxJS, don't know how to use/rewrite it using observableSpy, in either case all my test passes! 
And my code is simply composing an out string that "mimic" the "marble diagram" of the out$, because I need to test the result checking the stream of events that occurs, and the single value that can pass through mapfilter (odd numbers -> 2x) and which one are filtered (even number)

Reply:
Even if I use for semplicity a "sync" observable alle the value 1, 2, 3 are emitted as single number, notice I use in$= from([1,42,3]) that will produce an Observable with a "marble" representation like  this `>1-42-3-|` where 
`>` represents the start of the stream
`-` represents the time that sepatete one next from the subsequent
`|` is the complete event (end of the stream) 
`X` here not used, represents and error emission (that maybe end the stream)

Hope this will clarify even my test cases

Reply:
Yes I get the 'look like' marble test. But real marble test are hard to understand, so similar one are even harder. It's hard to get why you test it like that. 

My propose solution:
```ts
	it('given an observable >1-42-3| and a mapping function that double ONLY the odds value, then result is an observable of >"2"--"6"-| the intial even value (42) is not mapped and so filtered out', fakeAsync(() => {
		const in$ = of(1, 42, 3);
		const out$ = in$.pipe(
			mapFilter((n) => {
				if (n % 2) return String(n * 2);
				else return undefined; // explict return undefined to skipout (filter) some value from the out observable
			})
		);

		// let out = '>';
		out$.pipe(toArray()).subscribe((s) => {
			//s is a string
			expect(s).toEqual(['2', '6']);
		});
		tick();
		// out += '|';
		// expect(out).toEqual('>2-6-|');
	}));
```

or 
```ts
	it('given an observable >1-42-3| and a mapping function that double ONLY the odds value, then result is an observable of >"2"--"6"-| the intial even value (42) is not mapped and so filtered out', (done) => {
		const in$ = of(1, 42, 3);
		const out$ = in$.pipe(
			mapFilter((n) => {
				if (n % 2) return String(n * 2);
				else return undefined; // explict return undefined to skipout (filter) some value from the out observable
			})
		);

		// let out = '>';
		out$.pipe(toArray()).subscribe((s) => {
			//s is a string
			expect(s).toEqual(['2', '6']);
			done();
		});
		// out += '|';
		// expect(out).toEqual('>2-6-|');
	});
```

Reply:
Thanks for your suggested Test code, I've never known/used `toArray()` operator, quite un-usual one, but in this case can simplify the test expect. 
If you don't mind rewrite them I prefer to keep the "marble" representation...
PS: with **of**`([1,42,3])` the test will fail, you must use **from**`([1,42,3])`;

Reply:
I second the `observer-spy` approach. The tests work but they're hard to parse what's going on.

Reply:
Is this [observer-spy](https://www.npmjs.com/package/@hirez_io/observer-spy) you are referring?
Do I can install it? And have to update the jest.config to handle auto-unsubscribe?!

Please @nartc can you suggest me how to solve the problem with CI failing because I used the @ts-ignore to write the test for empty code-path that doesn’t return value, or I simply remove the test before committing…